### ISSUE TITLE
Brief exit message for KeyboardInterrupt

### ISSUE THREAD

**Issue by mk-pmb** [streamlink/streamlink#4188]
### Checklist

- [X] This is a feature request and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin requests](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22feature+request%22)

### Description

Thanks for maintaining streamlink! It entertains me a lot. :+1: :rocket:

I'd like an option to reduce the verbosity of `KeyboardInterrupt` errors, ideally to one line at most.

I run `streamlink 2.1.1+12.g3fdd208.dirty` on Ubuntu focal as the last step of a script. Sometimes a message from an early step of that script means that I need a manual prepare step before I should start watching, so I press Ctrl+c to abort. Most of the times, this causes streamlinks's python to print about 20 lines of Traceback before the message "KeyboardInterrupt", and a useless blank line after that. Thus I have to scroll up to see the messages from my script. It would be nice to have an easy way to opt-out of that stack trace and the blank line.

**Comment by bastimeyer** [comment]
> 2.1.1+12.g3fdd208.dirty

This version is outdated and modified, so no support for that here...

> KeyboardInterrupt

Most `KeyboardInterrupt` exceptions should be caught by `streamlink_cli`. Some of them are also caught in the `streamlink` module/package, so it depends on how you run Streamlink, which isn't clear from your description.

Could you please post the entire call stack of the `KeyboardInterrupt` exception you're seeing? This is important to know so that this can be fixed, as `SIGINT` is not supposed to print `KeyboardInterrupt` exceptions. However, what I can already tell you without looking at the code, any `KeyboardInterrupt` that gets raised before Streamlink gets properly initialized won't be caught and thus gets printed to `stderr`. This could be improved, but not fixed entirely AFAIA, unless we override the default signal behavior as the first step before doing anything else.

Any kind of error log gets written to `stderr` instead of `stdout`, so if you don't want to see that in your script, then there are ways to suppress that. `SIGTERM` should get interpreted the same as `SIGINT` (configured by `streamlink_cli`), but `SIGKILL` could also be an option if necessary, albeit a bit ugly.

**Comment by mk-pmb** [comment]
Thanks for the quick feedback! I already considered piping stderr trough sed to optimize it, but I still have hopes that a python tweak will be easier.

> any KeyboardInterrupt that gets raised before Streamlink gets properly initialized won't be caught

I checked with various timings and can confirm that this is my problem. When I give it enough time to initialize, the error message is just

> Interrupted! Exiting...

Detailed log for my outdated version:

<details>

```text
$ git reset --hard 3fdd208dbeeede10d80483c3e47b5f2752d1be7f
HEAD is now at 3fdd208d plugins.nicolive: fix proxy arguments (#3710)

$ rm -- src/streamlink/plugins/{mjunoon,ustvnow}.py \
      # ^-- Avoid errors about missing libraries

$ export PYTHONPATH="$PYTHONPATH:$PWD/src"

$ CAMP='python3 -m streamlink_cli twitch.tv/basecamp'

$ timeout --signal=SIGINT 1s $CAMP
[cli][info] Found matching plugin twitch for URL twitch.tv/basecamp
Interrupted! Exiting...

$ timeout --signal=SIGINT 0.7s $CAMP
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/streamlink-repo/src/streamlink_cli/__main__.py", line 3, in <module>
    main()
  File "/tmp/streamlink-repo/src/streamlink_cli/main.py", line 1032, in main
    setup_streamlink()
  File "/tmp/streamlink-repo/src/streamlink_cli/main.py", line 732, in setup_streamlink
    streamlink = Streamlink({"user-input-requester": ConsoleUserInputRequester(console)})
  File "/tmp/streamlink-repo/src/streamlink/session.py", line 79, in __init__
    self.load_builtin_plugins()
  File "/tmp/streamlink-repo/src/streamlink/session.py", line 445, in load_builtin_plugins
    self.load_plugins(plugins.__path__[0])
  File "/tmp/streamlink-repo/src/streamlink/session.py", line 458, in load_plugins
    mod = load_module(module_name, path)
  File "/tmp/streamlink-repo/src/streamlink/utils/__init__.py", line 24, in load_module
    spec.loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/tmp/streamlink-repo/src/streamlink/plugins/nbc.py", line 4, in <module>
    from streamlink.plugins.theplatform import ThePlatform
  File "/tmp/streamlink-repo/src/streamlink/plugins/theplatform.py", line 10, in <module>
    class ThePlatform(Plugin):
  File "/tmp/streamlink-repo/src/streamlink/plugins/theplatform.py", line 15, in ThePlatform
    release_re = re.compile(r'''tp:releaseUrl\s*=\s*"(.*?)"''')
  File "/usr/lib/python3.8/re.py", line 252, in compile
    return _compile(pattern, flags)
  File "/usr/lib/python3.8/re.py", line 304, in _compile
    p = sre_compile.compile(pattern, flags)
  File "/usr/lib/python3.8/sre_compile.py", line 764, in compile
    p = sre_parse.parse(p, flags)
  File "/usr/lib/python3.8/sre_parse.py", line 940, in parse
    source = Tokenizer(str)
  File "/usr/lib/python3.8/sre_parse.py", line 229, in __init__
    self.decoded_string = string
KeyboardInterrupt
```

</details>

I tried to check with the current version, too, but with

```text
HEAD is now at 33ccd2eb plugins.crunchyroll: add metadata attributes (#4185)
```

it fails with

> ModuleNotFoundError: No module named 'Crypto.Util.Padding'

which, last time I tried, was too cumbersome to obtain.

**Comment by bastimeyer** [comment]
The default behavior of the signal handlers can probably be changed until streamlink_cli is properly initialized. There shouldn't be a reason why we want a KeyboardInterrupt to be printed to stderr unless we're fully initialized. I will take a look at this later today.

> it fails with
> > ModuleNotFoundError: No module named 'Crypto.Util.Padding'

You are missing dependencies and have probably not installed Streamlink correctly. If a package is not available for your distro, then install it in a clean virtual environment or just use the appimage.

https://streamlink.github.io/install.html#pypi-package-and-source-code
https://streamlink.github.io/install.html#appimages

**Comment by mk-pmb** [comment]
> You are missing dependencies and have probably not installed Streamlink correctly.

Yes, I know. That's what I meant to convey with "too cumbersome to obtain". In the version I'm currently using, it's not a problem, becuase the dependency is only required by some plugins that I don't use anyway.
If you think it might help improve the KeyboardInterrupt issue, I could try and install the Crypto Padding, but I doubt it's related.

**Comment by bastimeyer** [comment]
If you install Streamlink in a virtual environment via python-pip, as shown in the linked documentation, then there won't be any missing dependencies unless you deliberately removed them again. The dependencies are pulled from pypi and not from your distro's package management, so this works everywhere and is not cumbersome. And yes, this is of course unrelated to the KeyboardInterrupt stuff, however, asking for support or providing log output when using an older and modified version is unnecessary.

As you can see above, I have already submitted a pull request which should fix the issue (#4190), or at least the vast majority of cases where sending SIGINT/SIGTERM would lead to a call stack being printed to stderr. As I mentioned in the PR, this can't be fixed entirely, but the solution should be good enough. You can test and verify it. Again, this is described in the documentation.
https://streamlink.github.io/install.html#pypi-package-and-source-code

```sh
virtualenv ~/venv/streamlink
source ~/venv/streamlink/bin/activate
pip install -U git+https://github.com/bastimeyer/streamlink.git@cli/override-default-signals
deactivate

~/venv/streamlink/bin/streamlink --loglevel debug
```

**Comment by gravyboat** [comment]
Locking since this is resolved and I don't want to see any follow up messages from other users about old versions.
